bin/annotate_test_failures

```diff
@@ -14,31 +18,102 @@ unless File.exist?(junit_path)
   exit 0
 end
 
+###################
+# Helper methods
+###################
+
+# Here's how report.junit XML files look like for failure vs success:
+#
+# <testcase classname='WordPressTest.MediaServiceTests' name='testDeletingLocalMediaThatDoesntExistInCoreData'>
+#   <failure message='Asynchronous wait failed: Exceeded timeout of 0.1 seconds, with unfulfilled expectations: &quot;The delete call succeeds even if the media object isn&apos;t saved.&quot;.'>&lt;unknown&gt;:0</failure>
+# </testcase>
+# <testcase classname='WordPressTest.MediaServiceTests' name='testDeletingLocalMediaThatDoesntExistInCoreData' time='0.145'/>
+#
+class TestFailure
+  attr_reader :classname, :name, :message, :details
+  attr_accessor :count
+
+  def initialize(node)
+    @classname = node['classname']
+    @name = node['name']
+    failure_node = node.elements['failure']
+    @message = failure_node['message']
+    @details = failure_node.text
+    @count = 1
+  end
+
+  def to_s
+    times = @count > 1 ? " (#{@count} times)" : ''
+    <<~ENTRY
+      <details><summary><tt>#{@name}</tt> in <tt>#{@classname}</tt>#{times}</summary>
+      #{@message}
+
+      ```
+      #{@details}
+      ```
+      </details>
+    ENTRY
+  end
+
+  def key
+    "#{@classname}.#{@name}"
+  end
+
+  def ultimately_succeeds?(parent_node:)
+    all_same_test_nodes = parent_node.get_elements("testcase[@classname='#{@classname}' and @name='#{@name}']")
+    !all_same_test_nodes.last.elements['failure'] # If last node found for that test doesn't have a <failure> child, then test ultimately succeeded.
+  end
+
+  def ==(other)
+    @classname == other.classname && @name == other.name && @message == other.message && @details == other.details
+  end
+end
+
+# Add a test_failure into a given array, or increment its count if it already exists
+#
+def record(test_failure, into:)
+  existing = into.find { |f| f == test_failure }
+  if existing.nil?
+    into.append(test_failure)
+  else
+    existing.count += 1
+  end
+end
+
+# Given a list of failures for a given step and state, update the corresponding annotation
+#
+def update_annotation(title, list, style, state)
+  annotation_context = "#{title} (#{state})"
+  if list.empty?
+    puts "No test #{state}. Removing any previous `#{annotation_context}` Buildkite annotation if any.\n\n"
+    system('buildkite-agent', 'annotation', 'remove', '--context', annotation_context)
+  else
+    tests_count = list.map(&:key).uniq.count # Count the number of tests that failed, even if some tests might have multiple assertion failures
+    puts "#{tests_count} test(s) #{state} (#{list.count} distinct assertion failures in total). Reporting them as a `#{annotation_context}` Buildkite #{style} annotation.\n\n"
```

Comment:
Looking at the CI annotations, I was going to suggest to use "test(s)" because "1 tests" reads odd. But I see you already addressed this 👍 

<img width="784" alt="Screenshot 2023-03-14 at 1 09 37 pm" src="https://user-images.githubusercontent.com/1218433/224874435-5d552f57-cb73-46d9-9839-a791fd3a66db.png">

Reply:
Yeah the test CI jobs on my dummy WPiOS test branch didn't pick up the change even if I retried the builds because of cache (since the plugin [was referenced in the `pipeline.yml` as `automattic/a8c-ci-toolkit#fix/annotate_test_failures`, so by branch name, not commit](https://github.com/wordpress-mobile/WordPress-iOS/blob/bcfad407146ab9077fb77a5a7e085340af776ff2/.buildkite/pipeline.yml#L5), and it didn't feel worth it to go all the way to invalidate cache or reference by commit and retry all the test builds just for getting that `(s)` change and update the screenshots 😄